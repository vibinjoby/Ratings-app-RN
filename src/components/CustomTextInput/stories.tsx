import React from 'react'
import { storiesOf } from '@storybook/react-native'
import CustomTextInput from './component'

storiesOf('Components', module).add('CustomTextInput', () => <CustomTextInput />)
