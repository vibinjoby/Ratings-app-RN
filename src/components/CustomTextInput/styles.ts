import { StyleSheet } from 'react-native'
import Colors from '../../utilities/colors'
import Typography from '../../utilities/typography'

export default StyleSheet.create({
  container: {
    flexDirection: 'row',
  },
  err: { color: Colors.red },
  placeholder: {
    fontSize: 16,
    zIndex: 999,
    position: 'absolute',
    marginTop: 5,
    fontWeight: '500',
    left: 0,
    top: -10,
    ...Typography.Caption1.regular,
  },
  textInput: {
    ...Typography.RegularFont,
    width: '100%',
    color: Colors.black,
    borderColor: Colors.textInputBorder,
    borderBottomWidth: 1,
    borderRadius: 4,
    paddingVertical: 15,
    paddingHorizontal: 3,
    fontSize: 16,
    marginBottom: 10,
    opacity: 0.8,
    marginTop: 10,
  },
  hidePwd: {
    position: 'absolute',
    right: 10,
    zIndex: 1,
    top: 15,
    width: 25,
    height: 25,
  },
  errTxt: {
    fontSize: 12,
    color: Colors.errorText,
    marginBottom: 5,
  },
})
