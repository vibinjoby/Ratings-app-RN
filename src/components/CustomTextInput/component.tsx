import React, { useState } from 'react'
import { View, Image, TextInput, TouchableWithoutFeedback, KeyboardType, Text } from 'react-native'
import * as Animatable from 'react-native-animatable'
import Colors from '../../utilities/colors'

import PasswordErrorTxt from '../PasswordErrorText'
import styles from './styles'

export interface CustomTextInputProps {
  testID: string
  onChangeText: (arg0: string) => void
  customStyles?: Record<string, unknown>
  touched: { [key: string]: boolean }
  errors: { [key: string]: string }
  name: string
  onPress?: () => void
  value?: string
  placeholderTxt?: string
  secureTextEntry?: boolean
  autoCapitalize?: boolean
  extraStyle?: Record<string, unknown>
  placeholderTextColor?: string
  keyboardType?: KeyboardType
  onBlur?: () => void
  onFocus?: () => void
}

const CustomTextInput: React.FC<CustomTextInputProps> = ({
  testID,
  onChangeText,
  customStyles,
  touched,
  errors,
  name,
  keyboardType,
  value,
  placeholderTxt,
  secureTextEntry,
  autoCapitalize,
  extraStyle,
}) => {
  const [isLabelVisible, setIsLabelVisible] = useState(false)
  const [isHidePwd, setIsHidePwd] = useState(secureTextEntry)
  const [inputValue, setInputValue] = useState('')
  const [isPasswordFocussed, setIsPasswordFocussed] = useState(false)

  return (
    <View style={customStyles}>
      <View style={styles.container}>
        {isLabelVisible && (
          <Animatable.Text
            style={[styles.placeholder, touched[name] && errors[name] ? styles.err : {}]}
          >
            {placeholderTxt}
          </Animatable.Text>
        )}
        <TextInput
          testID={testID}
          placeholder={placeholderTxt}
          returnKeyType={'done'}
          style={[
            styles.textInput,
            isLabelVisible && {
              borderColor: touched[name] && errors[name] ? Colors.red : Colors.enabledTxtInput,
            },
            extraStyle,
          ]}
          keyboardType={keyboardType}
          secureTextEntry={isHidePwd}
          autoCapitalize={!autoCapitalize ? 'none' : 'words'}
          onChangeText={(text: string) => {
            //If the text is typed move the placeholder to the top as label
            if (text && text !== '') setIsLabelVisible(true)
            else setIsLabelVisible(false)
            onChangeText && onChangeText(text)
            setInputValue(text)
            if (name && name === 'password') setIsPasswordFocussed(true)
            else setIsPasswordFocussed(false)
          }}
          value={value}
        />
      </View>
      {name === 'password' && (
        <TouchableWithoutFeedback onPress={() => setIsHidePwd(!isHidePwd)}>
          <Image
            resizeMode={'contain'}
            source={
              !isHidePwd ? require('../../assets/eye.png') : require('../../assets/eye-off.png')
            }
            style={styles.hidePwd}
          />
        </TouchableWithoutFeedback>
      )}
      {touched && errors[name] && name !== 'password' && (
        <Text style={styles.errTxt}>{errors[name]}</Text>
      )}
      {(isPasswordFocussed || touched[name]) && name === 'password' && (
        <PasswordErrorTxt inputValue={inputValue} />
      )}
    </View>
  )
}

export default CustomTextInput
