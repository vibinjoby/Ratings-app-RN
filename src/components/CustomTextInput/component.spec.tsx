import 'react-native'
import React from 'react'
import { render } from '@testing-library/react-native'

import CustomTextInput from './component'

describe('CustomTextInput', () => {
  it('should render', () => {
    const { toJSON } = render(
      <CustomTextInput
        testID="items"
        key={''}
        onBlur={() => console.log('blurred')}
        onFocus={() => console.log('validate form')}
        onChangeText={() => console.log('onchange text')}
        touched={{ key: false }}
        name={'textinput'}
        errors={{ key: 'errors' }}
      />,
    )
    expect(toJSON()).toMatchSnapshot()
  })
})
