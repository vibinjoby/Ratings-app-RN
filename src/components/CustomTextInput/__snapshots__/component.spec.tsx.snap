// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomTextInput should render 1`] = `
<View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      autoCapitalize="none"
      onChangeText={[Function]}
      returnKeyType="done"
      style={
        Array [
          Object {
            "borderBottomWidth": 1,
            "borderColor": "rgba(30, 33, 76,0.1)",
            "borderRadius": 4,
            "color": "#181725",
            "fontFamily": "Poppins-Regular",
            "fontSize": 16,
            "marginBottom": 10,
            "marginTop": 10,
            "opacity": 0.8,
            "paddingHorizontal": 3,
            "paddingVertical": 15,
            "width": "100%",
          },
          false,
          undefined,
        ]
      }
      testID="items"
    />
  </View>
</View>
`;
